File: articles/genomics-expansion-of-biological-knowledge/c/genomics_kernel.c
/*
 * Compact genomics kernel in C.
 */

#include <math.h>
#include <stdio.h>

double expected_heterozygosity(double p) {
    return 2.0 * p * (1.0 - p);
}

double fst_style(double p1, double p2) {
    double pbar = (p1 + p2) / 2.0;
    double ht = 2.0 * pbar * (1.0 - pbar);
    double hs = (2.0 * p1 * (1.0 - p1) + 2.0 * p2 * (1.0 - p2)) / 2.0;

    if (ht <= 0.0) {
        return 0.0;
    }

    return (ht - hs) / ht;
}

double jukes_cantor(double d) {
    if (d >= 0.75) {
        return NAN;
    }

    return -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * d);
}

int main(void) {
    double p = 0.8;
    double q = 1.0 - p;
    double log2fc = log((160.0 + 1.0) / (100.0 + 1.0)) / log(2.0);

    printf("AA=%.6f\n", p * p);
    printf("Aa=%.6f\n", 2.0 * p * q);
    printf("aa=%.6f\n", q * q);
    printf("expected_heterozygosity=%.6f\n", expected_heterozygosity(p));
    printf("fst_style=%.6f\n", fst_style(0.40, 0.75));
    printf("jukes_cantor=%.6f\n", jukes_cantor(0.15));
    printf("log2fc=%.6f\n", log2fc);

    return 0;
}

File: articles/genomics-expansion-of-biological-knowledge/fortran/genomics_kernel.f90
! Compact genomics numerical kernel in Fortran.

program genomics_kernel
  implicit none

  real :: p
  real :: q
  real :: he
  real :: p1
  real :: p2
  real :: pbar
  real :: ht
  real :: hs
  real :: fst
  real :: d
  real :: jc
  real :: treated
  real :: control
  real :: log2fc

  p = 0.8
  q = 1.0 - p
  he = 2.0 * p * q

  print *, "AA:", p ** 2
  print *, "Aa:", 2.0 * p * q
  print *, "aa:", q ** 2
  print *, "Expected heterozygosity:", he

  p1 = 0.40
  p2 = 0.75
  pbar = (p1 + p2) / 2.0
  ht = 2.0 * pbar * (1.0 - pbar)
  hs = (2.0 * p1 * (1.0 - p1) + 2.0 * p2 * (1.0 - p2)) / 2.0
  fst = (ht - hs) / ht

  print *, "FST-style value:", fst

  d = 0.15
  jc = -(3.0 / 4.0) * log(1.0 - (4.0 / 3.0) * d)

  print *, "Jukes-Cantor distance:", jc

  treated = 160.0
  control = 100.0
  log2fc = log((treated + 1.0) / (control + 1.0)) / log(2.0)

  print *, "log2FC:", log2fc
end program genomics_kernel
